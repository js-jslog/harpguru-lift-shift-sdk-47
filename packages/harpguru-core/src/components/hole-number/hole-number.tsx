import { View } from 'react-native'
import React from 'react'
import type { ReactElement } from 'react'

import { TextWithoutOSScale } from '../text-without-os-scale'
import type { Coord } from '../../types'

import { useStyles } from './hooks'

export enum HoleNumberIds {
  One = '1',
  Two = '2',
  Three = '3',
  Four = '4',
  Five = '5',
  Six = '6',
  Seven = '7',
  Eight = '8',
  Nine = '9',
  Ten = '10',
  Eleven = '11',
  Twelve = '12',
  Thirteen = '13',
  Fourteen = '14',
  Fifteen = '15',
  Sixteen = '16',
}

type HoleNumberProps = {
  readonly xCoord: Coord
}

export const HoleNumber = ({ xCoord }: HoleNumberProps): ReactElement => {
  const styles = useStyles()
  const holeNumber: HoleNumberIds = `${xCoord + 1}` as HoleNumberIds

  return (
    <View style={styles.cell}>
      <TextWithoutOSScale style={styles.text}>{holeNumber}</TextWithoutOSScale>
    </View>
  )
}
