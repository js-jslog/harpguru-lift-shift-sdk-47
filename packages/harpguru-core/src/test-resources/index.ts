export {
  inactiveCellsHarpStrata,
  activeCellsHarpStrata,
  chromaticHarpStrata,
  buildMockUseGlobalImplementation,
} from './test-resources'
