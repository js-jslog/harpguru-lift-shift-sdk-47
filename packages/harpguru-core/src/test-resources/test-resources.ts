import { getHarpStrata } from 'harpstrata'
import type { HarpStrata, HarpStrataProps } from 'harpstrata'
import {
  TuningIds,
  PitchIds,
  DegreeIds,
  PozitionIds,
  ValvingIds,
} from 'harpparts'

import {
  reduceFullMatrixToViewableMatrix,
  reduceLayoutFactsToDynamicSizes,
  reduceToStaticSizes,
  reduceMatrixToLayoutFacts,
} from '../utils'
import { DisplayModes, ExperienceModes, FlushChannels } from '../types'
import type { ColumnBounds } from '../types'

const allActiveDegrees = [
  DegreeIds.Root,
  DegreeIds.Flat2,
  DegreeIds.Second,
  DegreeIds.Flat3,
  DegreeIds.Third,
  DegreeIds.Fourth,
  DegreeIds.Flat5,
  DegreeIds.Fifth,
  DegreeIds.Flat6,
  DegreeIds.Sixth,
  DegreeIds.Flat7,
  DegreeIds.Seventh,
]

const baseHarpStrataProps: HarpStrataProps = {
  tuningId: TuningIds.MajorDiatonic,
  valvingId: ValvingIds.NotValved,
  pozitionId: PozitionIds.Second,
  harpKeyId: PitchIds.C,
  activeIds: [] as const,
}
const inactiveCellsHarpStrataProps = baseHarpStrataProps
const activeCellsHarpStrataProps = {
  ...baseHarpStrataProps,
  activeIds: allActiveDegrees,
}
const chromaticHarpStrataProps = {
  ...baseHarpStrataProps,
  tuningId: TuningIds.SixteenHoleChromatic,
}

export const inactiveCellsHarpStrata = getHarpStrata(
  inactiveCellsHarpStrataProps
)
export const activeCellsHarpStrata = getHarpStrata(activeCellsHarpStrataProps)
export const chromaticHarpStrata = getHarpStrata(chromaticHarpStrataProps)

type BuildMockUseGlobalImplementationProps = {
  readonly sourceHarpStrata?: HarpStrata
  readonly sourceColumnBounds?: ColumnBounds
  readonly experienceMode?: ExperienceModes
  readonly displayMode?: DisplayModes
  readonly bufferedActivityToggles?: ReadonlyArray<DegreeIds>
  readonly fragmentHarpFaceByOctaves?: boolean
  readonly flushChannel?: FlushChannels
  readonly activeQuizDegrees?: ReadonlyArray<DegreeIds>
}

type MockGlobalImplementation = (arg0: string) => unknown[] | undefined

export const buildMockUseGlobalImplementation = ({
  sourceHarpStrata = activeCellsHarpStrata,
  sourceColumnBounds = 'FIT',
  experienceMode = ExperienceModes.Explore,
  displayMode = DisplayModes.Degree,
  bufferedActivityToggles = [],
  fragmentHarpFaceByOctaves = true,
  flushChannel = FlushChannels.Regular,
  activeQuizDegrees = allActiveDegrees,
}: BuildMockUseGlobalImplementationProps): MockGlobalImplementation => {
  const mockGlobal = (stateItem: string) => {
    if (stateItem === 'activeDegreeMatrix')
      return [sourceHarpStrata.degreeMatrix]
    if (stateItem === 'activePitchMatrix') return [sourceHarpStrata.pitchMatrix]
    if (stateItem === 'activeInteractionMatrix')
      return [sourceHarpStrata.apparatus.interactionMatrix]
    if (stateItem === 'activeDegreeIds')
      return [sourceHarpStrata.activeDegreeIds]
    if (stateItem === 'activePitchIds') return [sourceHarpStrata.activePitchIds]
    if (stateItem === 'tuningId') return [sourceHarpStrata.apparatus.tuningId]
    if (stateItem === 'valvingId') return [sourceHarpStrata.apparatus.valvingId]
    if (stateItem === 'harpKeyId') return [sourceHarpStrata.harpKeyId]
    if (stateItem === 'pozitionId') return [sourceHarpStrata.pozitionId]
    if (stateItem === 'rootPitchId') return [sourceHarpStrata.rootPitchId]

    if (stateItem === 'sourceColumnBounds')
      return [sourceColumnBounds, jest.fn()]
    if (stateItem === 'columnBounds') return [sourceColumnBounds]
    const viewableDegreeMatrix = reduceFullMatrixToViewableMatrix(
      { harpface1: [[]] },
      sourceHarpStrata.degreeMatrix,
      sourceHarpStrata.apparatus.interactionMatrix,
      sourceColumnBounds
    )
    const viewablePitchMatrix = reduceFullMatrixToViewableMatrix(
      { harpface1: [[]] },
      sourceHarpStrata.pitchMatrix,
      sourceHarpStrata.apparatus.interactionMatrix,
      sourceColumnBounds
    )
    const viewableInteractionMatrix = reduceFullMatrixToViewableMatrix(
      { harpface1: [[]] },
      sourceHarpStrata.apparatus.interactionMatrix,
      sourceHarpStrata.apparatus.interactionMatrix,
      sourceColumnBounds
    )
    if (stateItem === 'viewableDegreeMatrix') return [viewableDegreeMatrix]
    if (stateItem === 'viewablePitchMatrix') return [viewablePitchMatrix]
    if (stateItem === 'viewableInteractionMatrix')
      return [viewableInteractionMatrix]

    const fullLayoutFacts = reduceMatrixToLayoutFacts(
      {
        harpface1: {
          harpfaceColumns: 0,
          harpfaceRows: 0,
        },
      },
      sourceHarpStrata.apparatus.interactionMatrix
    )
    if (stateItem === 'fullLayoutFacts') return [fullLayoutFacts]
    const layoutFacts = reduceMatrixToLayoutFacts(
      {
        harpface1: {
          harpfaceColumns: 0,
          harpfaceRows: 0,
        },
      },
      viewableInteractionMatrix
    )
    if (stateItem === 'layoutFacts') return [layoutFacts]

    const dynamicSizes = reduceLayoutFactsToDynamicSizes(undefined, {
      fullLayoutFacts,
      layoutFacts,
    })
    if (stateItem === 'dynamicSizes') return [dynamicSizes]

    const staticSizes = reduceToStaticSizes(undefined)
    if (stateItem === 'staticSizes') return [staticSizes]

    if (stateItem === 'activeExperienceMode') return [experienceMode]
    if (stateItem === 'activeDisplayMode') return [displayMode]
    if (stateItem === 'bufferedActivityToggles')
      return [bufferedActivityToggles]
    if (stateItem === 'fragmentHarpFaceByOctaves')
      return [fragmentHarpFaceByOctaves]
    if (stateItem === 'flushChannel') return [flushChannel]
    if (stateItem === 'activeQuizDegrees') return [activeQuizDegrees]
    return undefined
  }
  return mockGlobal
}
