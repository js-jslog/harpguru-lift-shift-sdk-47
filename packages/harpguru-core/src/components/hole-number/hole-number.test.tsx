import { useGlobal } from 'reactn'
import React from 'react'
import { render } from '@testing-library/react-native'

import { buildMockUseGlobalImplementation } from '../../test-resources'

import { HoleNumber, HoleNumberIds } from './hole-number'

jest.mock('reactn')
const mockUseGlobal = useGlobal as jest.Mock
mockUseGlobal.mockImplementation(buildMockUseGlobalImplementation({}))

test('HoleNumber renders a dom element with the expected value included', () => {
  const { getByText } = render(<HoleNumber xCoord={0} />)
  expect(getByText(HoleNumberIds.One)).toBeTruthy()
})

test('A snapshot of HoleNumber', () => {
  const { toJSON } = render(<HoleNumber xCoord={0} />)
  expect(toJSON()).toMatchSnapshot()
})
