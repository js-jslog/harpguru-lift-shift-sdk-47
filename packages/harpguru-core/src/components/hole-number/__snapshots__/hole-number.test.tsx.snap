// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A snapshot of HoleNumber 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "height": 22.034250207026385,
      "justifyContent": "center",
      "top": -11.017125103513193,
      "width": 57.6864220478633,
    }
  }
>
  <Text
    allowFontScaling={false}
    style={
      Object {
        "color": "#555",
        "fontSize": 22.034250207026385,
      }
    }
  >
    1
  </Text>
</View>
`;
